src/replication.c

```diff
@@ -3505,240 +3715,100 @@ void syncWithPrimary(connection *conn) {
      * may find that the socket is in error state. */
     if (connGetState(conn) != CONN_STATE_CONNECTED) {
         serverLog(LL_WARNING, "Error condition on socket for SYNC: %s", connGetLastError(conn));
-        goto error;
+        syncWithPrimaryHandleError(&conn);
+        return;
     }
-
+    switch (server.repl_state) {
     /* Send a PING to check the primary is able to reply without errors. */
-    if (server.repl_state == REPL_STATE_CONNECTING) {
-        serverLog(LL_NOTICE, "Non blocking connect for SYNC fired the event.");
-        /* Delete the writable event so that the readable event remains
-         * registered and we can wait for the PONG reply. */
-        connSetReadHandler(conn, syncWithPrimary);
-        connSetWriteHandler(conn, NULL);
+    case REPL_STATE_CONNECTING:
+        if (syncWithPrimaryHandleConnectingState(conn, &err) == C_ERR) {
+            syncWithPrimaryHandleWriteError(&conn, &err);
+            return;
+        }
         server.repl_state = REPL_STATE_RECEIVE_PING_REPLY;
-        /* Send the PING, don't check for errors at all, we have the timeout
-         * that will take care about this. */
-        err = sendCommand(conn, "PING", NULL);
-        if (err) goto write_error;
         return;
-    }
-
     /* Receive the PONG command. */
-    if (server.repl_state == REPL_STATE_RECEIVE_PING_REPLY) {
-        err = receiveSynchronousResponse(conn);
-
-        /* The primary did not reply */
-        if (err == NULL) goto no_response_error;
-
-        /* We accept only two replies as valid, a positive +PONG reply
-         * (we just check for "+") or an authentication error.
-         * Note that older versions of Redis OSS replied with "operation not
-         * permitted" instead of using a proper error code, so we test
-         * both. */
-        if (err[0] != '+' && strncmp(err, "-NOAUTH", 7) != 0 && strncmp(err, "-NOPERM", 7) != 0 &&
-            strncmp(err, "-ERR operation not permitted", 28) != 0) {
-            serverLog(LL_WARNING, "Error reply to PING from primary: '%s'", err);
-            sdsfree(err);
-            goto error;
-        } else {
-            serverLog(LL_NOTICE, "Primary replied to PING, replication can continue...");
+    case REPL_STATE_RECEIVE_PING_REPLY:
+        if (syncWithPrimaryHandleReceivePingReplyState(conn, &err) == C_ERR) {
+            syncWithPrimaryHandleError(&conn);
```

Comment:
are we leaking `err`?

```suggestion
            syncWithPrimaryHandleWriteError(&conn, &err);
```

Reply:
No leak - err is now freed inside `syncWithPrimaryHandleReceivePingReplyState`

Reply:
>err is now freed inside syncWithPrimaryHandleReceivePingReplyState

This is an implementation detail. we should go by the function contract. I would say either we free `err` here or we should change `syncWithPrimaryHandleReceivePingReplyState` to not return `err`. either way would work for me.

Reply:
Good idea - thinking about this more, we can remove `syncWithPrimaryHandleWriteError` completely, since all it does on top of `syncWithPrimaryHandleError` is log a line, which we can move to inside the handlers. Then none of the `syncWithPrimaryHandler`s will have to take `err` as an argument - they will allocate and free it internally.  What do you think?

Reply:
Yeah removing `syncWithPrimaryHandleWriteError` and consolidating on `syncWithPrimaryHandleError` is good. That being said, I would still like us to fix `syncWithPrimaryHandleReceivePingReplyState` to either return an `err` string that needs to be freed by the caller or not return an `err` at all.

Reply:
Done